File: backend/lite/scaffold.py
from node import ConstantNode, LookupNode, RequestNode, OutputNode
from value import SPRef
from omegadb import OmegaDB
from detach import unapplyPSP
from regen import applyPSP

class Scaffold(object):
  def __init__(self,setsOfPNodes=None,regenCounts=None,absorbing=None,aaa=None,border=None,lkernels=None,brush=None):
    self.setsOfPNodes = setsOfPNodes if setsOfPNodes else [] # [Set Node]
    self.regenCounts = regenCounts if regenCounts else {} # {Node:Int}
    self.absorbing = absorbing if absorbing else set() # Set Node
    self.aaa = aaa if aaa else set() # Set Node
    self.border = border if border else [] # [[Node]]
    self.lkernels = lkernels if lkernels else {} # {Node:LKernel}
    self.brush = brush if brush else set() # Set Node

  def getPrincipalNodes(self):
    # Return a list so that repeated traversals have the same order
    return [n for n in set.union(*self.setsOfPNodes)]
  def getRegenCount(self,node): return self.regenCounts[node]
  def incrementRegenCount(self,node): self.regenCounts[node] += 1
  def decrementRegenCount(self,node): self.regenCounts[node] -= 1
  def isResampling(self,node): return node in self.regenCounts
  def isAbsorbing(self,node): return node in self.absorbing
  def isAAA(self,node): return node in self.aaa
  def hasLKernel(self,node): return node in self.lkernels
  def getLKernel(self,node): return self.lkernels[node]
  def getPNode(self):
    assert len(self.setsOfPNodes) == 1
    pnodes = []
    for pnode in self.setsOfPNodes[0]: pnodes.append(pnode)
    assert len(pnodes) == 1
    return pnodes[0]
  def isBrush(self, node): return node in self.brush

  def show(self):
    print "---Scaffold---"
    print "# pnodes: " + str(len(self.getPrincipalNodes()))
    print "# absorbing nodes: " + str(len(self.absorbing))
    print "# aaa nodes: " + str(len(self.aaa))
    print "border lengths: " + str([len(segment) for segment in self.border])
    print "# lkernels: " + str(len(self.lkernels))

  def showMore(self):
    print "---Scaffold---"
    print "pnodes: " + str(self.getPrincipalNodes())
    print "absorbing nodes: " + str(self.absorbing)
    print "aaa nodes: " + str(self.aaa)
    print "borders: " + str(self.border)
    print "lkernels: " + str(self.lkernels)

"""
When subsampled_mh is called, there will exist broken deterministic relationships in the trace.
For example, in the program:
[assume mu (normal 0 1)]
[assume x (lambda () (normal mu 1))]
[observe (x) 0.1]
[observe (x) 0.2]
...
N observations
...
After mu is updated. If a subsampled scaffold only include the first n observations,
the lookup nodes for mu in the remaining N-n observations will have a stale value.
At a second call to infer, these inconsistency may cause a problem.

updateValuesAtScaffold updates all the nodes in a newly constructed scaffold by
calling updateValueAtNode for each node. The main idea is that:
Assume all the random output nodes have the latest values and the problem is in
the nodes with determnistic dependency on their parents.
For every node that cannot absorb, update the value of all its parents, and then
unapplyPSP and apply PSP. For a request node, compare its old Request with the new
request. If they are different, unevalRequests and evalRequests to generate new
ESRs. When a new ESR is generated, return True. Otherwise return false.
"""
#FIXME(Yutian): At regen phase in proposal, value should also be updated. This will
#involve changing the regen method.
def updateValuesAtScaffold(trace,scaffold,updatedNodes):
  # for every node in the scaffold (resampling, brush, border) update values.
  # return true if any node gets a new value.
  hasNewEsr = False
  #for node in scaffold.brush:
  #  hasNewEsr |= updateValueAtNode(trace, scaffold, node, updatedNodes)
  for node in scaffold.regenCounts:
    hasNewEsr |= updateValueAtNode(trace, scaffold, node, updatedNodes)
  #for borderList in scaffold.border:
  #  for node in borderList:
  #    hasNewEsr |= updateValueAtNode(trace, scaffold, node, updatedNodes)

  return hasNewEsr

def updateValueAtNode(trace, scaffold, node, updatedNodes):
  # Return True if a new value is assigned.
  if node in updatedNodes:
    return False

  # Strong assumption! Only consider resampling nodes in the scaffold.
  if not scaffold.isResampling(node):
    return False

  hasNewEsr = False
  if isinstance(node, ConstantNode):
    return False
  elif isinstance(node, LookupNode):
    hasNewEsr |= updateValueAtNode(trace, scaffold, node.sourceNode, updatedNodes)
    trace.setValueAt(node, trace.valueAt(node.sourceNode))
  elif isinstance(node, RequestNode):
    return False
  else: # OutputNode.
    # Assume SPRef and AAA nodes are always updated.
    psp = trace.pspAt(node)
    if isinstance(trace.valueAt(node), SPRef) or psp.childrenCanAAA():
      pass

    canAbsorb = True
    for parent in trace.parentsAt(node):
      if not psp.canAbsorb(trace, node, parent):
        hasNewEsr |= updateValueAtNode(trace, scaffold, parent, updatedNodes)
        canAbsorb = False

    # Assume AAA node is always updated
    if not canAbsorb:
      #DEBUG
      #scaffold = Scaffold()
      #omegaDB = OmegaDB()
      #oldValue = trace.valueAt(node)
      # DEBUG
      #unapplyPSP(trace, node, scaffold, omegaDB)
      #applyPSP(trace,node,scaffold,False,omegaDB,{})
      updateValue(trace, node)
      #print "Regen output node", oldValue, trace.valueAt(node), psp

  #hasNewEsr = False
  #if isinstance(node, ConstantNode):
  #  pass
  #elif isinstance(node, LookupNode):
  #  hasNewEsr |= updateValueAtNode(trace, scaffold, node.sourceNode, updatedNodes)
  #  trace.setValueAt(node, trace.valueAt(node.sourceNode))
  #else: # Application Node.
  #  psp = trace.pspAt(node)
  #  canAbsorb = True
  #  for parent in trace.parentsAt(node):
  #    canAbsorb &= psp.canAbsorb(trace, node, parent)
  #  # Assume AAA node is always updated
  #  if not canAbsorb and not psp.childrenCanAAA():
  #    # psp can not absorb for all the paretns
  #    for parent in trace.definiteParentsAt(node):
  #      hasNewEsr |= updateValueAtNode(trace, scaffold, parent, updatedNodes)

  #    # Update esrParent value after update the value of the request node.
  #    if isinstance(node, OutputNode):
  #      for esrParent in trace.esrParentsAt(node):
  #        hasNewEsr |= updateValueAtNode(trace, scaffold, esrParent, updatedNodes)

  #    # Regen value.
  #    if isinstance(node, RequestNode):
  #      # FIXME: only checked the value of esrs (in string)
  #      # FIXME: psp is unapplied before unevalRequests, and the args of psp may not be the same as they were used for applyPSP. Problem for AAA?
  #      #DEBUG
  #      #print "Regen request node", psp

  #      #scaffold = Scaffold()
  #      #omegaDB = OmegaDB()
  #      #old_request = trace.valueAt(node)
  #      ##DEBUG
  #      #spf = trace.spFamiliesAt(node)
  #      #for esr in old_request.esrs:
  #      #  if esr.id not in spf.families:
  #      #    print "strange"

  #      #unapplyPSP(trace, node, scaffold, omegaDB)
  #      #applyPSP(trace,node,scaffold,False,omegaDB,{})
  #      #new_request = trace.valueAt(node)

  #      ##DEBUG
  #      #spf = trace.spFamiliesAt(node)
  #      #for esr in new_request.esrs:
  #      #  if esr.id not in spf.families:
  #      #    print "strange"

  #      #eq_request = len(old_request.esrs) == len(new_request.esrs)
  #      #if eq_request:
  #      #  for i in range(len(old_request.esrs)):
  #      #    if not old_request.esrs[i].id == new_request.esrs[i].id:
  #      #      eq_request = False
  #      #      break
  #      #if not eq_request:
  #      #  #DEBUG
  #      #  old_request_str = str(old_request.esrs)
  #      #  new_request_str = str(new_request.esrs)
  #      #  print "not equal request", old_request_str, new_request_str
  #      #  trace.setValueAt(node, old_request)
  #      #  unevalRequests(trace, node, scaffold, omegaDB, compute_gradient)
  #      #  trace.setValueAt(node, new_request)
  #      #  evalRequests(trace,requestNode,scaffold,shouldRestore,omegaDB,gradients)
  #      #  hasNewEsr |= True
  #      pass
  #    else: # OutputNode
  #      if not isinstance(trace.valueAt(node), SPRef):
  #        #DEBUG
  #        scaffold = Scaffold()
  #        omegaDB = OmegaDB()
  #        #oldValue = trace.valueAt(node)
  #        # DEBUG
  #        #unapplyPSP(trace, node, scaffold, omegaDB)
  #        #applyPSP(trace,node,scaffold,False,omegaDB,{})
  #        updateValue(trace, node, scaffold, omegaDB)
  #        #print "Regen output node", oldValue, trace.valueAt(node), psp

  updatedNodes.add(node)
  return hasNewEsr

# DEBUG, for cprofiler only
def updateValue(trace, node):
  scaffold = Scaffold()
  omegaDB = OmegaDB()
  unapplyPSP(trace, node, scaffold, omegaDB)
  applyPSP(trace,node,scaffold,False,omegaDB,{})


def constructScaffold(trace,setsOfPNodes,useDeltaKernels = False, deltaKernelArgs = None, updateValue = False, updatedNodes = None):
  if updatedNodes is None:
    updatedNodes = set()
  while True:
    cDRG,cAbsorbing,cAAA = set(),set(),set()
    indexAssignments = {}
    assert isinstance(setsOfPNodes,list)
    for i in range(len(setsOfPNodes)):
      assert isinstance(setsOfPNodes[i],set)
      extendCandidateScaffold(trace,setsOfPNodes[i],cDRG,cAbsorbing,cAAA,indexAssignments,i)

    brush = findBrush(trace,cDRG)
    drg,absorbing,aaa = removeBrush(cDRG,cAbsorbing,cAAA,brush)
    border = findBorder(trace,drg,absorbing,aaa)
    regenCounts = computeRegenCounts(trace,drg,absorbing,aaa,border,brush)
    lkernels = loadKernels(trace,drg,aaa,useDeltaKernels,deltaKernelArgs)
    borderSequence = assignBorderSequnce(border,indexAssignments,len(setsOfPNodes))
    scaffold = Scaffold(setsOfPNodes,regenCounts,absorbing,aaa,borderSequence,lkernels,brush)
    if not updateValue or not updateValuesAtScaffold(trace,scaffold,updatedNodes):
      break
  return scaffold

def addResamplingNode(trace,drg,absorbing,aaa,q,node,indexAssignments,i):
  if node in absorbing: absorbing.remove(node)
  if node in aaa: aaa.remove(node)
  drg.add(node)
  q.extend([(n,False,node) for n in trace.childrenAt(node)])
  indexAssignments[node] = i

def addAbsorbingNode(drg,absorbing,aaa,node,indexAssignments,i):
  assert not node in drg
  assert not node in aaa
  absorbing.add(node)
  indexAssignments[node] = i

def addAAANode(drg,aaa,absorbing,node,indexAssignments,i):
  if node in absorbing: absorbing.remove(node)
  drg.add(node)
  aaa.add(node)
  indexAssignments[node] = i


def extendCandidateScaffold(trace,pnodes,drg,absorbing,aaa,indexAssignments,i):
  q = [(pnode,True,None) for pnode in pnodes]

  while q:
    node,isPrincipal,parentNode = q.pop()
    if node in drg and not node in aaa:
      addResamplingNode(trace,drg,absorbing,aaa,q,node,indexAssignments,i)
    elif isinstance(node,LookupNode) or node.operatorNode in drg:
      addResamplingNode(trace,drg,absorbing,aaa,q,node,indexAssignments,i)
    # TODO temporary: once we put all uncollapsed AAA procs into AEKernels, this line won't be necessary
    elif node in aaa:
      addAAANode(drg,aaa,absorbing,node,indexAssignments,i)      
    elif (not isPrincipal) and trace.pspAt(node).canAbsorb(trace,node,parentNode):
      addAbsorbingNode(drg,absorbing,aaa,node,indexAssignments,i)
    elif trace.pspAt(node).childrenCanAAA(): 
      addAAANode(drg,aaa,absorbing,node,indexAssignments,i)
    else: 
      addResamplingNode(trace,drg,absorbing,aaa,q,node,indexAssignments,i)


def findBrush(trace,cDRG):
  disableCounts = {}
  disabledRequests = set()
  brush = set()
  for node in cDRG:
    if isinstance(node,RequestNode):
      disableRequests(trace,node,disableCounts,disabledRequests,brush)
  return brush

def disableRequests(trace,node,disableCounts,disabledRequests,brush):
  if node in disabledRequests: return
  disabledRequests.add(node)
  for esrParent in trace.esrParentsAt(node.outputNode):
    if not esrParent in disableCounts: disableCounts[esrParent] = 0
    disableCounts[esrParent] += 1
    if disableCounts[esrParent] == esrParent.numRequests:
      disableFamily(trace,esrParent,disableCounts,disabledRequests,brush)

def disableFamily(trace,node,disableCounts,disabledRequests,brush):
  if node in brush: return
  brush.add(node)
  if isinstance(node,OutputNode):
    brush.add(node.requestNode)
    disableRequests(trace,node.requestNode,disableCounts,disabledRequests,brush)
    disableFamily(trace,node.operatorNode,disableCounts,disabledRequests,brush)
    for operandNode in node.operandNodes: 
      disableFamily(trace,operandNode,disableCounts,disabledRequests,brush)

def removeBrush(cDRG,cAbsorbing,cAAA,brush):
  drg = cDRG - brush
  absorbing = cAbsorbing - brush
  aaa = cAAA - brush
  assert aaa.issubset(drg)
  assert not drg.intersection(absorbing)
  return drg,absorbing,aaa

def hasChildInAorD(trace,drg,absorbing,node):
  kids = trace.childrenAt(node)
  return kids.intersection(drg) or kids.intersection(absorbing)

def findBorder(trace,drg,absorbing,aaa):
  border = absorbing.union(aaa)
  for node in drg - aaa:
    if not hasChildInAorD(trace,drg,absorbing,node): border.add(node)
  return border

def maybeIncrementAAARegenCount(trace,regenCounts,aaa,node):
  value = trace.valueAt(node)
  if isinstance(value,SPRef) and value.makerNode in aaa: 
    regenCounts[value.makerNode] += 1

def computeRegenCounts(trace,drg,absorbing,aaa,border,brush):
  regenCounts = {}
  for node in drg:
    if node in aaa:
      regenCounts[node] = 1 # will be added to shortly
    elif node in border:
      regenCounts[node] = len(trace.childrenAt(node)) + 1
    else:
      regenCounts[node] = len(trace.childrenAt(node))
  
  if aaa:
    for node in drg.union(absorbing):
      for parent in trace.parentsAt(node):
        maybeIncrementAAARegenCount(trace,regenCounts,aaa,parent)

    for node in brush:
      if isinstance(node,OutputNode):
        for esrParent in trace.esrParentsAt(node):
          maybeIncrementAAARegenCount(trace,regenCounts,aaa,esrParent)
      elif isinstance(node,LookupNode):
        maybeIncrementAAARegenCount(trace,regenCounts,aaa,node.sourceNode)

  return regenCounts

def loadKernels(trace,drg,aaa,useDeltaKernels,deltaKernelArgs):
  lkernels = { node : trace.pspAt(node).getAAALKernel() for node in aaa}
  if useDeltaKernels:
    for node in drg - aaa:
      if not isinstance(node,OutputNode): continue
      if node.operatorNode in drg: continue
      for o in node.operandNodes:
        if o in drg: continue
      if trace.pspAt(node).hasDeltaKernel(): lkernels[node] = trace.pspAt(node).getDeltaKernel(deltaKernelArgs)
  return lkernels

def assignBorderSequnce(border,indexAssignments,numIndices):
  borderSequence = [[] for _ in range(numIndices)]
  for node in border:
    borderSequence[indexAssignments[node]].append(node)
  return borderSequence

File: backend/new_cxx/src/scaffold.cxx
#include "scaffold.h"
#include "node.h"
#include "concrete_trace.h"
#include "sp.h"
#include <algorithm>
#include <boost/lexical_cast.hpp>

set<Node *> Scaffold::getPrincipalNodes() 
{ 
  assert(setsOfPNodes.size() == 1);
  return setsOfPNodes[0];
}

int Scaffold::getRegenCount(Node * node) 
{ 
  assert(regenCounts.count(node));
  assert(regenCounts[node] >= 0);
  return regenCounts[node]; 
}

void Scaffold::incRegenCount(Node * node) 
{
  assert(regenCounts.count(node));
  regenCounts[node]++; 
}

void Scaffold::decRegenCount(Node * node) 
{ 
  assert(regenCounts.count(node));
  regenCounts[node]--;
  assert(regenCounts[node] >= 0);
}

bool Scaffold::isResampling(Node * node) { return regenCounts.count(node); }
bool Scaffold::isAbsorbing(Node * node) { return absorbing.count(node); }
bool Scaffold::isAAA(Node * node) { return aaa.count(node); }
bool Scaffold::hasLKernel(Node * node) { return lkernels.count(node); }
shared_ptr<LKernel> Scaffold::getLKernel(Node * node) 
{
  assert(lkernels.count(node));
  return lkernels[node];
}

string Scaffold::showSizes()
{
  string p = "(";
  return p + boost::lexical_cast<string>(regenCounts.size()) + "," + boost::lexical_cast<string>(absorbing.size()) + "," + boost::lexical_cast<string>(aaa.size()) + "," + boost::lexical_cast<string>(border[0].size()) + ")";
}


shared_ptr<Scaffold> constructScaffold(ConcreteTrace * trace,const vector<set<Node*> > & setsOfPNodes,bool useDeltaKernels)
{
  set<Node*> cDRG;
  set<Node*> cAbsorbing;
  set<Node*> cAAA;
  map<Node*,int> indexAssignments;

  for (size_t i = 0; i < setsOfPNodes.size(); ++i)
  {
    extendCandidateScaffold(trace,setsOfPNodes[i],cDRG,cAbsorbing,cAAA,indexAssignments,i);
  }
  set<Node*> brush = findBrush(trace,cDRG,cAbsorbing,cAAA);
  tuple<set<Node*>,set<Node*>,set<Node*> > coreScaffold  = removeBrush(cDRG,cAbsorbing,cAAA,brush);
  set<Node*> drg = boost::get<0>(coreScaffold);
  set<Node*> absorbing = boost::get<1>(coreScaffold);
  set<Node*> aaa = boost::get<2>(coreScaffold);

  set<Node*> border = findBorder(trace,drg,absorbing,aaa);
  map<Node*,int> regenCounts = computeRegenCounts(trace,drg,absorbing,aaa,border,brush);
  map<Node*,shared_ptr<LKernel> > lkernels = loadKernels(trace,drg,aaa,false);
  vector<vector<Node *> > borderSequence = assignBorderSequnce(border,indexAssignments,setsOfPNodes.size());
  return shared_ptr<Scaffold>(new Scaffold(setsOfPNodes,regenCounts,absorbing,aaa,borderSequence,lkernels));
}


void addResamplingNode(ConcreteTrace * trace,
                       set<Node*> & cDRG,
                       set<Node*> & cAbsorbing,
                       set<Node*> & cAAA,
                       queue<tuple<Node*,bool,Node*> > & q,
                       Node * node,
                       map<Node*,int> & indexAssignments,
                       int i)
{
  if (cAbsorbing.count(node)) { cAbsorbing.erase(node); }
  if (cAAA.count(node)) { cAAA.erase(node); }
  cDRG.insert(node);
  set<Node*> children = trace->getChildren(node);
  for (set<Node*>::iterator childIter = children.begin();
       childIter != children.end();
       ++childIter)
  {
    q.push(make_tuple(*childIter,false,node));
  }
  indexAssignments[node] = i;
}

void addAbsorbingNode(set<Node*> & cDRG,
                      set<Node*> & cAbsorbing,
                      set<Node*> & cAAA,
                      Node * node,
                      map<Node*,int> & indexAssignments,
                      int i) 
{
  assert(!cDRG.count(node));
  assert(!cAAA.count(node));
  cAbsorbing.insert(node);
  indexAssignments[node] = i;
}

void addAAANode(set<Node*> & cDRG,
                set<Node*> & cAbsorbing,
                set<Node*> & cAAA,
                Node * node,
                map<Node*,int> & indexAssignments,
                int i) 
{
  if (cAbsorbing.count(node)) { cAbsorbing.erase(node); }
  cDRG.insert(node);
  cAAA.insert(node);
  indexAssignments[node] = i;
}

void extendCandidateScaffold(ConcreteTrace * trace,
                             const set<Node*> & pnodes,
                             set<Node*> & cDRG,
                             set<Node*> & cAbsorbing,
                             set<Node*> & cAAA,
                             map<Node*,int> & indexAssignments,
                             int i) 
{
  queue<tuple<Node*,bool,Node*> > q;
  for (set<Node*>::iterator pnodeIter = pnodes.begin();
       pnodeIter != pnodes.end();
       ++pnodeIter)
  {
    Node * nullNode = NULL;
    q.push(make_tuple(*pnodeIter,true,nullNode));
  }

  while (!q.empty())
  {
    tuple<Node*,bool,Node*> qElem = q.front();
    Node * node = boost::get<0>(qElem);
    bool isPrincipal = boost::get<1>(qElem);
    Node * parentNode = boost::get<2>(qElem);
    q.pop();

    if (cDRG.count(node) && !cAAA.count(node)) { }
    else if (dynamic_cast<LookupNode*>(node))
    {
      addResamplingNode(trace,cDRG,cAbsorbing,cAAA,q,node,indexAssignments,i);
    }
    else
    {
      ApplicationNode * appNode = dynamic_cast<ApplicationNode*>(node);
      assert(appNode);
      shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(appNode))->getPSP(appNode);
      if (cDRG.count(appNode->operatorNode))
      {
        addResamplingNode(trace,cDRG,cAbsorbing,cAAA,q,node,indexAssignments,i);
      }
      else if (cAAA.count(appNode)) { }
      else if (!isPrincipal && psp->canAbsorb(trace,appNode,parentNode))
      {
        addAbsorbingNode(cDRG,cAbsorbing,cAAA,appNode,indexAssignments,i);
      }
      else if (psp->childrenCanAAA())
      {
        addAAANode(cDRG,cAbsorbing,cAAA,node,indexAssignments,i);
      }
      else 
      { 
        addResamplingNode(trace,cDRG,cAbsorbing,cAAA,q,node,indexAssignments,i); 
      }
    }
  }
}

set<Node*> findBrush(ConcreteTrace * trace,
                     set<Node*> & cDRG,
                     set<Node*> & cAbsorbing,
                     set<Node*> & cAAA) 
{
  map<RootOfFamily,int> disableCounts;
  set<RequestNode*> disabledRequests;
  set<Node*> brush;
  for (set<Node*>::iterator drgIter = cDRG.begin();
       drgIter != cDRG.end();
       ++drgIter)
  {
    RequestNode * requestNode = dynamic_cast<RequestNode*>(*drgIter);
    if (requestNode) { disableRequests(trace,requestNode,disableCounts,disabledRequests,brush); }
  }
  return brush;
}

void disableRequests(ConcreteTrace * trace,
                     RequestNode * node,
                     map<RootOfFamily,int> & disableCounts,
                     set<RequestNode*> & disabledRequests,
                     set<Node*> & brush)
{
  if (disabledRequests.count(node)) { return; }
  disabledRequests.insert(node);
  vector<RootOfFamily> esrRoots = trace->getESRParents(node->outputNode);
  for (size_t i = 0; i < esrRoots.size(); ++i) 
  { 
    RootOfFamily esrRoot = esrRoots[i];
    disableCounts[esrRoot] += 1;
    if (disableCounts[esrRoot] == trace->getNumRequests(esrRoot))
    {
      disableFamily(trace,esrRoot.get(),disableCounts,disabledRequests,brush);
    }
  }
}


void disableFamily(ConcreteTrace * trace,
                   Node * node,
                   map<RootOfFamily,int> & disableCounts,
                   set<RequestNode*> & disabledRequests,
                   set<Node*> & brush) 
{
  if (brush.count(node)){ return; }
  brush.insert(node);
  OutputNode * outputNode = dynamic_cast<OutputNode*>(node);
  if (outputNode)
  {
    brush.insert(outputNode->requestNode);
    disableRequests(trace,outputNode->requestNode,disableCounts,disabledRequests,brush);
    disableFamily(trace,outputNode->operatorNode,disableCounts,disabledRequests,brush);
    for (size_t i = 0; i < outputNode->operandNodes.size(); ++i)
    {
      disableFamily(trace,outputNode->operandNodes[i],disableCounts,disabledRequests,brush);
    }
  }
}


tuple<set<Node*>,set<Node*>,set<Node*> > removeBrush(set<Node*> & cDRG,
                                                     set<Node*> & cAbsorbing,
                                                     set<Node*> & cAAA,
                                                     set<Node*> & brush) 
{
  set<Node*> drg;
  set<Node*> absorbing;
  set<Node*> aaa;

  // TODO confirm that this works
  std::set_difference(cDRG.begin(),cDRG.end(),brush.begin(),brush.end(), 
                      std::inserter(drg,drg.begin()));

  std::set_difference(cAbsorbing.begin(),cAbsorbing.end(),brush.begin(),brush.end(), 
                      std::inserter(absorbing,absorbing.begin()));

  std::set_difference(cAAA.begin(),cAAA.end(),brush.begin(),brush.end(), 
                      std::inserter(aaa,aaa.begin()));

  /* DEBUG */
  // assert(std::includes(aaa.begin(),aaa.end(),drg.begin(),drg.end()));
  // set<Node*> intersection;
  // set_intersection(drg.begin(),drg.end(),absorbing.begin(),absorbing.end(), std::inserter(intersection,intersection.begin()));
  // assert(intersection.empty());
  /* END DEBUG */

  return make_tuple(drg,absorbing,aaa);
}

bool hasChildInAorD(ConcreteTrace * trace,
                    set<Node*> & drg,
                    set<Node*> & absorbing,
                    Node * node) 
{
  set<Node *> children = trace->getChildren(node);
  for (set<Node*>::iterator childIter = children.begin();
       childIter != children.end();
       ++childIter)
  {
    if (drg.count(*childIter) || absorbing.count(*childIter)) { return true; }
  }
  return false;
}

set<Node*> findBorder(ConcreteTrace * trace,
                      set<Node*> & drg,
                      set<Node*> & absorbing,
                      set<Node*> & aaa) 
{
  set<Node*> border;
  std::set_union(absorbing.begin(),absorbing.end(),aaa.begin(),aaa.end(), 
                 std::inserter(border,border.begin()));
  for (set<Node*>::iterator drgIter = drg.begin(); drgIter != drg.end(); ++drgIter)
  {
    if (!hasChildInAorD(trace,drg,absorbing,*drgIter)) { border.insert(*drgIter); }
  }
  return border;
}

void maybeIncrementAAARegenCount(ConcreteTrace * trace,
                                 map<Node*,int> & regenCounts,
                                 set<Node*> & aaa,
                                 Node * node) 
{
  shared_ptr<VentureSPRef> spRef = dynamic_pointer_cast<VentureSPRef>(trace->getValue(node));
  if (spRef && aaa.count(spRef->makerNode)) { regenCounts[spRef->makerNode] += 1; }
}

map<Node*,int> computeRegenCounts(ConcreteTrace * trace,
                                  set<Node*> & drg,
                                  set<Node*> & absorbing,
                                  set<Node*> & aaa,
                                  set<Node*> & border,
                                  set<Node*> & brush) 
{
  map<Node*,int> regenCounts;
  for (set<Node*>::iterator drgIter = drg.begin(); drgIter != drg.end(); ++drgIter)
  {
    if (aaa.count(*drgIter)) { regenCounts[*drgIter] = 1; } // will be added to shortly
    else if (border.count(*drgIter)) { regenCounts[*drgIter] = trace->getChildren(*drgIter).size() + 1; }
    else { regenCounts[*drgIter] = trace->getChildren(*drgIter).size(); }
  }

  if (!aaa.empty())
  {
    for (set<Node*>::iterator drgIter = drg.begin(); drgIter != drg.end(); ++drgIter)
    {
      vector<Node*> parents = trace->getParents(*drgIter);
      for (size_t i = 0; i < parents.size(); ++i) { maybeIncrementAAARegenCount(trace,regenCounts,aaa,parents[i]); }
    }

    for (set<Node*>::iterator absorbingIter = absorbing.begin(); absorbingIter != absorbing.end(); ++absorbingIter)
    {
      vector<Node*> parents = trace->getParents(*absorbingIter);
      for (size_t i = 0; i < parents.size(); ++i) { maybeIncrementAAARegenCount(trace,regenCounts,aaa,parents[i]); }
    }
    
    for (set<Node*>::iterator brushIter = brush.begin(); brushIter != brush.end(); ++brushIter)
    {
      LookupNode * lookupNode = dynamic_cast<LookupNode*>(*brushIter);
      OutputNode * outputNode = dynamic_cast<OutputNode*>(*brushIter);
      if (outputNode) 
      {
        vector<RootOfFamily> esrParents = trace->getESRParents(*brushIter);
        for (size_t i = 0; i < esrParents.size(); ++i) { maybeIncrementAAARegenCount(trace,regenCounts,aaa,esrParents[i].get()); }
      }
      if (lookupNode)
      {
        maybeIncrementAAARegenCount(trace,regenCounts,aaa,lookupNode->sourceNode);
      }
    }
  }
  return regenCounts;
}

map<Node*,shared_ptr<LKernel> > loadKernels(ConcreteTrace * trace,
                                            set<Node*> & drg,
                                            set<Node*> & aaa,
                                            bool useDeltaKernels)
{ 
  map<Node*,shared_ptr<LKernel> > lkernels;
  for (set<Node*>::iterator aaaIter = aaa.begin(); aaaIter != aaa.end(); ++aaaIter)
  {
    OutputNode * outputNode = dynamic_cast<OutputNode*>(*aaaIter);
    assert(outputNode);
    shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(outputNode))->getPSP(outputNode);
    lkernels[*aaaIter] = psp->getAAALKernel();
  }

  if (useDeltaKernels)
    {
      for (set<Node*>::iterator drgIter = drg.begin(); drgIter != drg.end(); ++drgIter)
        {
          OutputNode * outputNode = dynamic_cast<OutputNode*>(*drgIter);
          if (!outputNode) { continue; }
          if (drg.count(outputNode->operatorNode)) { continue; }
          if (aaa.count(outputNode)) { continue; }
          shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(outputNode))->getPSP(outputNode);
          if (psp->hasDeltaKernel()) { lkernels[outputNode] = psp->getDeltaKernel(); }
        }
    }
  return lkernels;
}


vector<vector<Node *> > assignBorderSequnce(set<Node*> & border,
                                            map<Node*,int> & indexAssignments,
                                            int numIndices) 
{
  vector<vector<Node *> > borderSequence(numIndices);
  for (set<Node*>::iterator borderIter = border.begin(); borderIter != border.end(); ++borderIter)
  {
    borderSequence[indexAssignments[*borderIter]].push_back(*borderIter);
  }
  return borderSequence;
}
